chainer/iterators/multithread_iterator.py

```diff
@@ -0,0 +1,217 @@
+        self._data_queue = six.moves.queue.Queue()
+        self._cnt = 0
+
+        self._workers = []
+
+        self._finalized = finalized
+
+        self._pool = pool.ThreadPool(self.n_threads)
+        self._next = None
+
+    def _invoke_prefetch(self):
+        assert self._next is None
+        n = len(self.dataset)
+        i = self._pushed_position
+        if i is None:  # first iteration
+            i = self.current_position
+
+        order = self._order
+        args = []
+        for _ in six.moves.range(self.batch_size):
+            if i >= n:
+                if not self._repeat:
+                    break
+                i = 0
+                if order is not None:
+                    # We cannot shuffle the order directly here, since the
+                    # iterator may be serialized before the prefetched data are
+                    # consumed by the user, in which case an inconsistency
+                    # appears.
+                    order = order.copy()
+                    numpy.random.shuffle(order)
+            index = i if order is None else order[i]
+            args.append(index)
+            i += 1
+
+        dataset = self.dataset
+
+        def _read(index):
+            return dataset[index]
+
+        self._next = self._pool.map_async(_read, args)
+        self._prefetch_order = order  # Temporarily store the shuffled order.
+        self._pushed_position = i
+
+    def _get(self):
+        n = len(self.dataset)
+        i = self.current_position
+
+        batch = []
+        next = self._next
+        while not next.ready():
+            next.wait(0.5)
+
+        for data in next.get():
+            batch.append(data)
+            i += 1
+            if i >= n:
+                self.epoch += 1
+                self.is_new_epoch = True
+                i = 0
+                if not self._repeat:
+                    break
+        self._next = None
+
+        self.current_position = i
+        # Eventually overwrite the (possibly shuffled) order.
+        self._order = self._prefetch_order
+        return batch
+
+    def reset(self):
+        if getattr(self, 'current_position', 0) != 0:
+            raise NotImplementedError(
+                'Reset of MultithreadIterator in the middle of a epoch is '
+                'currently not supported.')
+        if getattr(self, 'epoch', 0) != 0 and self._repeat:
+            raise NotImplementedError(
+                'Reset of repeating MultithreadIterator is currently not '
+                'supported.')
+        if getattr(self, '_finalized', None) is not None and \
+                self._finalized.is_set():
+            raise NotImplementedError(
+                'Reset of finalized MultithreadIterator is currently not '
+                'supported.')
+
+        self.current_position = 0
+        self.epoch = 0
+        self.is_new_epoch = False
+
+        # use -1 instead of None internally.
+        self._previous_epoch_detail = -1.
```

Comment:
Do we have to hold `-1`? How about holding `None` and converting values just before/after serialization?